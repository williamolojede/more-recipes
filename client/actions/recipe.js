import instance from '../config/axios';

import { setFetching, unsetFetching } from './fetching';
import { showNotification } from './notification';

import * as actionTypes from './types';

const receiveRecipes = (type, res) => ({
  type,
  recipes: res.data.recipes,
  pagination: res.data.pagination
});

const recieveSingleRecipe = (type, recipe) => ({
  type,
  recipe
});

const fetchRecipesError = (err, dispatch, type) => {
  dispatch({
    type,
    errorMessage: err.response.data.message
  });
  dispatch(unsetFetching());
};


export const resetSingleRecipe = () => ({
  type: actionTypes.RESET_SINGLE_RECIPE
});

export const fetchSingleRecipe = id => (dispatch) => {
  dispatch(setFetching());
  return instance.get(`/recipes/${id}`)
    .then((res) => {
      dispatch(recieveSingleRecipe(actionTypes.RECEIVE_SINGLE_RECIPE, res.data.recipe));
      dispatch(unsetFetching());
    })
    .catch(err => fetchRecipesError(err, dispatch, actionTypes.FETCH_SINGLE_RECIPE_ERROR));
};

export const fetchPersonalRecipes = ({ userId, page, limit }) => (dispatch) => {
  dispatch(setFetching());
  return instance.get(`/users/${userId}/recipes?page=${page}&limit=${limit}`)
    .then((res) => {
      dispatch(receiveRecipes(actionTypes.RECEIVE_PERSONAL_RECIPES, res));
      dispatch(unsetFetching());
    })
    .catch(err => fetchRecipesError(err, dispatch, actionTypes.FETCH_PERSONAL_RECIPES_ERROR));
};

export const fetchFavoriteRecipes = ({ userId, page, limit }) => (dispatch) => {
  dispatch(setFetching());
  return instance.get(`/users/${userId}/favorites?page=${page}&limit=${limit}`)
    .then((res) => {
      dispatch(receiveRecipes(actionTypes.RECEIVE_FAVORITE_RECIPES, res));
      dispatch(unsetFetching());
    })
    .catch(err => fetchRecipesError(err, dispatch, actionTypes.FETCH_FAVORITE_RECIPES_ERROR));
};

export const fetchTopRecipes = (page, limit) => (dispatch) => {
  dispatch(setFetching());
  return instance.get(`/recipes?sort=upvotes&order=descending&page=${page}&limit=${limit}`)
    .then((res) => {
      dispatch(receiveRecipes(actionTypes.RECEIVE_TOP_RATED_RECIPES, res));
      dispatch(unsetFetching());
    })
    .catch(err => fetchRecipesError(err, dispatch, actionTypes.FETCH_TOP_RECIPES_ERROR));
};

export const vote = (dir, id) => dispatch => instance.post(`/recipes/${id}/vote-${dir}`)
  .then((res) => {
    dispatch(recieveSingleRecipe(actionTypes.RECEIVE_SINGLE_RECIPE, res.data.recipe));
    dispatch(showNotification(res.data.message));
  })
  .catch((err) => {
    dispatch(showNotification(err.response.data.message));
  });


export const favorite = id => dispatch => instance.post(`/recipes/${id}/favorite`)
  .then((res) => {
    dispatch(recieveSingleRecipe(actionTypes.RECEIVE_SINGLE_RECIPE, res.data.recipe));
    dispatch(showNotification(res.data.message));

    // Return true to confirm delete
    // occurred used in favorite recipe page
    return true;
  })
  .catch((err) => {
    dispatch(showNotification(err.response.data.message));
  });

export const addRecipe = recipe => dispatch =>
  instance.post('/recipes', { recipe })
    .then((res) => {
      dispatch(showNotification(res.data.message));
    })
    .catch((err) => {
      dispatch(showNotification(err.response.data.message));
    });

export const updateRecipe = recipe => (dispatch) => {
  const {
    id,
    name,
    description,
    img_url,
    ingredients,
    instructions
  } = recipe;
  const update = {
    name,
    description,
    img_url,
    ingredients,
    instructions
  };
  return instance.put(`/recipes/${id}`, { update })
    .then((res) => {
      dispatch(showNotification(res.data.message));
    })
    .catch((err) => {
      dispatch(showNotification(err.response.data.message));
    });
};

export const deleteRecipe = id => (dispatch) => {
  dispatch(setFetching());
  return instance.delete(`recipes/${id}`)
    .then((res) => {
      dispatch(showNotification(res.data.message));
      dispatch(unsetFetching());

      // Return true to confirm delete
      // occurred used in personal recipe page
      return true;
    })
    .catch((err) => {
      dispatch(showNotification(err.response.data.message));
      dispatch(unsetFetching());
    });
};
