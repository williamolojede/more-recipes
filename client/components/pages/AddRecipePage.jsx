import React from 'react';

import ManageRecipes from '../shared/ManageRecipes';
import AddRecipeForm from '../shared/AddRecipeForm';

const AddRecipePage = props => (
  <ManageRecipes>
    <AddRecipeForm isEditMode={false} {...props} />
  </ManageRecipes>
);
export default AddRecipePage;
