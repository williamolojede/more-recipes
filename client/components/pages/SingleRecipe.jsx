/* eslint camelcase: 0 */

import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';


import {
  fetchSingleRecipe,
  vote,
  favorite,
  resetSingleRecipe
} from '../../actions/recipe';


import Preloader from '../shared/Preloader.jsx';
import ConnectedSiteNav from '../shared/SiteNav.jsx';
import SiteFooter from '../shared/SiteFooter.jsx';
import IngredientsList from '../shared/IngredientsList.jsx';
import InstructionsList from '../shared/InstructionsList.jsx';
import Reviews from '../shared/Reviews';
import RecipeStats from '../shared/RecipeStats.jsx';
import Notification from '../shared/Notification.jsx';

import { recipePropTypes, currentUserPropTypes } from '../../config/proptypes';

class SingleRecipe extends Component {
  componentDidMount() {
    this.props.dispatch(fetchSingleRecipe(this.props.match.params.id));
  }

  componentWillUnmount() {
    this.props.dispatch(resetSingleRecipe());
  }

  vote = (dir) => {
    this.props.dispatch(vote(dir, this.props.recipe.id));
  }

  favorite = () => {
    this.props.dispatch(favorite(this.props.recipe.id));
  }

  renderBody = () => {
    const { isFetching } = this.props;
    const {
      name,
      description,
      img_url,
      ingredients,
      instructions,
      upVoteCount,
      downVoteCount,
      favoriteCount,
      viewCount
    } = this.props.recipe;

    const stats = {
      upVoteCount,
      downVoteCount,
      favoriteCount,
      viewCount
    };

    // the default value of recipe.name => ''
    // which means the recipe hasnt been received yet
    if (isFetching && this.props.recipe.name === '') {
      return (<Preloader />);
    } else if (this.props.errorMessage) {
      return (
        <div className="recipe-not-found">
          <h1>
            <span role="img" aria-label="sad">ðŸ˜¢</span> Recipe Not Found
          </h1>
        </div>
      );
    }

    return (
      <main>
        <div className="container">
          <div className="row">
            <div className="col s12 m5">
              <img src={img_url} alt={name} />
            </div>
            <div className="col s12 m7">
              <h4 className="recipe__name">{name}</h4>
              <p className="recipe__description">{description}</p>
              <div className="row recipe__stats">
                <RecipeStats stats={stats} vote={this.vote} favorite={this.favorite} />
                <a href="#write-review" className="write-review">
                Write Review
                </a>
              </div>
            </div>
          </div>

          <IngredientsList ingredients={ingredients} />
          <InstructionsList instructions={instructions} />
          {
            name !== '' && <Reviews recipeId={parseInt(this.props.match.params.id, 10)} />
          }
        </div>
      </main>
    );
  };

  render() {
    return (
      <div className="page page__recipe">
        <header className="site-header">
          <ConnectedSiteNav currentUser={this.props.currentUser} />
        </header>
        {
          this.renderBody()
        }
        <Notification notification={this.props.notification} dispatch={this.props.dispatch} />
        <SiteFooter />
      </div>
    );
  }
}

SingleRecipe.propTypes = {
  dispatch: PropTypes.func.isRequired,

  ...recipePropTypes,
  ...currentUserPropTypes,

  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired,
    })
  }).isRequired,

  notification: PropTypes.string.isRequired
};

export const mapStateToProps = ({
  singleRecipe: { recipe, errorMessage },
  isFetching,
  auth: { currentUser },
  notification
}) => ({
  recipe,
  errorMessage,
  isFetching,
  notification,
  currentUser,
});

export { SingleRecipe as PureSingleRecipe };
export default connect(mapStateToProps)(SingleRecipe);
