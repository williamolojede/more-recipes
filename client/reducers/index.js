/**
 * Root Reducer
 */

import { combineReducers } from 'redux';
import { routerReducer } from 'react-router-redux';

// REDUCERS
import auth from './auth';
import isFetching from './isFetching';
import recipes from './recipes';

const rootReducer = combineReducers({
  auth,
  isFetching,
  recipes,
  router: routerReducer
});


export default rootReducer;
