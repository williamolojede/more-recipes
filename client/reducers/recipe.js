/**
 * Reducer for a Single Recipe
 */
import {
  RECEIVE_SINGLE_RECIPE,
  FETCH_SINGLE_RECIPE_ERROR,
  RESET_SINGLE_RECIPE
} from '../actions/types';
import defaultConfig from '../config/default';

export const initialState = {
  recipe: {
    id: 1,
    name: '',
    description: '',
    img_url: defaultConfig.recipeImgUrl,
    ingredients: [],
    instructions: [],
    upVoteCount: 0,
    downVoteCount: 0,
    favoriteCount: 0,
    viewCount: 0
  },
  errorMessage: ''
};

const singleRecipe = (state = initialState, { type, recipe, errorMessage }) => {
  switch (type) {
    case RECEIVE_SINGLE_RECIPE:
      return { ...state, ...{ recipe } };
    case FETCH_SINGLE_RECIPE_ERROR:
      return { ...state, ...{ errorMessage } };
    case RESET_SINGLE_RECIPE:
      return initialState;
    default:
      return state;
  }
};

export default singleRecipe;
