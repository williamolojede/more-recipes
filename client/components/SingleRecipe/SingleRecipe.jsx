/* eslint camelcase: 0 */

import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';


import { fetchSingleRecipes, vote, favorite } from '../../actions/recipe';

import Preloader from '../shared/Preloader.jsx';
import ConnectedSiteNav from '../shared/SiteNav.jsx';
import SiteFooter from '../shared/SiteFooter.jsx';
import IngredientsList from '../shared/IngredientsList.jsx';
import InstructionsList from '../shared/InstructionsList.jsx';
import Reviews from '../shared/Reviews/Reviews.jsx';
import RecipeStats from '../shared/RecipeStats.jsx';
import Notification from '../shared/Notification.jsx';

import { recipePropTypes, currentUserPropTypes } from '../../config/proptypes';
import { currentUserDefaultProps } from '../../config/defaultPropTypes';

class SingleRecipe extends Component {
  componentDidMount() {
    this.props.dispatch(fetchSingleRecipes(this.props.match.params.id));
  }

  vote = (dir) => {
    this.props.dispatch(vote(dir, this.props.recipe.id));
  }

  favorite = () => {
    this.props.dispatch(favorite(this.props.recipe.id));
  }

  render() {
    if (Object.keys(this.props.recipe).length === 0) return (<Preloader />);

    const {
      id,
      name,
      description,
      img_url,
      ingredients,
      instructions,
      reviews,
      upVoteCount,
      downVoteCount,
      favoriteCount,
      viewCount
    } = this.props.recipe;

    const stats = {
      upVoteCount,
      downVoteCount,
      favoriteCount,
      viewCount
    };


    return (
      <div className="page page__recipe">
        <header className="site-header">
          <ConnectedSiteNav currentUser={this.props.currentUser} />
        </header>
        <main>
          <div className="container">
            <div className="row">
              <div className="col s12 m5">
                <img src={img_url} alt={name} />
              </div>
              <div className="col s12 m7">
                <h4 className="recipe__name">{name}</h4>
                <p className="recipe__description">{description}</p>
                <div className="row recipe__stats">
                  <RecipeStats stats={stats} vote={this.vote} favorite={this.favorite} />
                  <a href="#write-review" className="write-review">
                      Write Review
                  </a>
                </div>
              </div>
            </div>

            <IngredientsList ingredients={ingredients} />
            <InstructionsList instructions={instructions} />
            <Reviews recipeId={id} />
          </div>
        </main>

        <Notification notification={this.props.notification} dispatch={this.props.dispatch} />
        <SiteFooter />
      </div>
    );
  }
}

SingleRecipe.defaultProps = {
  ...currentUserDefaultProps
};

SingleRecipe.propTypes = {
  dispatch: PropTypes.func.isRequired,

  ...recipePropTypes,
  ...currentUserPropTypes,

  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired,
    })
  }).isRequired,

  notification: PropTypes.string.isRequired
};

const mapStateToProps = ({
  recipe,
  isFetching,
  auth: { currentUser },
  notification
}) => ({
  recipe,
  isFetching,
  notification,
  currentUser,
});

export default connect(mapStateToProps)(SingleRecipe);
