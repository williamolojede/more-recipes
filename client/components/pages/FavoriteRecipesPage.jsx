import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';
import swal from 'sweetalert';


import decodeToken from '../../utils/decodeToken';

import ManageRecipes from '../shared/ManageRecipes';
import Preloader from '../shared/Preloader';
import RecipeCardShort from '../shared/RecipeCard/RecipeCardShort';

import { fetchFavoriteRecipes, favorite } from '../../actions/recipe';

class FavoriteRecipesPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      recipes: [],
      currentPage: 1,
      limit: 12,
      last: 1,
    };

    this.payload = {
      userId: decodeToken().id,
      page: this.state.currentPage,
      limit: this.state.limit
    };
  }

  componentWillMount() {
    const payload = {
      userId: decodeToken().id,
      page: this.state.currentPage,
      limit: this.state.limit
    };
    this.props.dispatch(fetchFavoriteRecipes(payload));
  }

  componentWillReceiveProps(newProps) {
    if (newProps.recipes) {
      const {
        recipes,
        pagination
      } = newProps;
      this.setState({
        recipes,
        last: pagination.last
      });
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevState.currentPage !== this.state.currentPage) {
      const payload = {
        userId: decodeToken().id,
        page: this.state.currentPage,
        limit: this.state.limit
      };
      this.props.dispatch(fetchFavoriteRecipes(payload));
    }
  }

  handlePageClick = ({ selected }) => {
    this.setState({
      currentPage: selected + 1
    });
  }

  removeRecipe = (id) => {
    swal({
      title: 'Remove recipe from your favorite List?',
      icon: 'warning',
      buttons: true,
      dangerMode: true,
    })
      .then((willDelete) => {
        if (willDelete) {
          return this.props.dispatch(favorite(id));
        }
        return willDelete;
      })
      .then((willDelete) => {
        if (!willDelete) return;
        const payload = {
          userId: decodeToken().id,
          page: (this.state.recipes.length === 1 && this.state.last !== 1) ?
            this.state.currentPage - 1 : this.state.currentPage,
          limit: this.state.limit
        };
        this.props.dispatch(fetchFavoriteRecipes(payload));
      });
  }

  renderBody = () => {
    const { isFetching } = this.props;

    if (isFetching) {
      return (<Preloader />);
    } else if (!isFetching && this.state.recipes.length === 0) {
      return (
        <div className="no-favorites">
          <span>Welcome to your favorites!</span>
          <span>Tap <i className="mdi mdi-heart" /> to save any recipe you like,</span>
          <span>and it&apos;ll show up here.</span>
        </div>
      );
    }
    return (
      <ul className="row recipes-list">
        {
          this.state.recipes.map(({ id, name }) => (<RecipeCardShort
            id={id}
            name={name}
            removeRecipe={this.removeRecipe}
            key={id}
            type="favorite"
          />))
        }
      </ul>
    );
  }

  render() {
    return (
      <ManageRecipes>
        <div className="container page-manage__recipes">

          {
            this.renderBody()
          }
          {
            (this.state.last > 1 && this.state.recipes.length !== 0) &&
            <footer>
              <ReactPaginate
                previousLabel={<i className="material-icons">chevron_left</i>}
                nextLabel={<i className="material-icons">chevron_right</i>}
                pageCount={this.state.last}
                marginPagesDisplayed={2}
                pageClassName="waves-effect"
                pageRangeDisplayed={4}
                onPageChange={this.handlePageClick}
                activeClassName="active"
                containerClassName={`pagination ${this.props.isFetching ? 'is-fetching' : ''}`}
              />
            </footer>
          }

        </div>
      </ManageRecipes>
    );
  }
}

FavoriteRecipesPage.propTypes = {
  dispatch: PropTypes.func.isRequired,
  isFetching: PropTypes.bool.isRequired,
};

const mapStateToProps = ({
  favoriteRecipes: { recipes, pagination },
  isFetching,
}) => ({
  recipes,
  pagination,
  isFetching
});

export { FavoriteRecipesPage as PureFavoriteRecipesPage };
export default connect(mapStateToProps)(FavoriteRecipesPage);
