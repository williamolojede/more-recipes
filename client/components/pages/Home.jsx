import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ReactPaginate from 'react-paginate';

import { fetchTopRecipes } from '../../actions/recipe';

import TopRatedRecipeList from '../shared/TopRatedRecipeList';
import Preloader from '../shared/Preloader';
import ConnectedSiteNav from '../shared/SiteNav';
import SiteFooter from '../shared/SiteFooter';
import Search from '../shared/Search';

import { currentUserPropTypes } from '../../config/proptypes';

class Home extends Component {
  constructor(props) {
    super(props);
    this.state = {
      recipes: [],
      currentPage: 1,
      limit: 6,
      last: 1,
      isSearchActive: false
    };
  }

  componentDidMount() {
    this.props.dispatch(fetchTopRecipes(this.state.currentPage, this.state.limit));
  }

  componentWillReceiveProps(newProps) {
    if (newProps.recipes) {
      const {
        recipes,
        pagination
      } = newProps;
      this.setState({
        recipes,
        last: pagination.last
      });
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevState.currentPage !== this.state.currentPage) {
      this.props.dispatch(fetchTopRecipes(this.state.currentPage, this.state.limit));
    }
  }

  toggleSearch =() => {
    this.setState({
      isSearchActive: !this.state.isSearchActive
    });
  }

  handlePageClick = ({ selected }) => {
    this.setState({
      currentPage: selected + 1
    });
  }

  renderBody = () => {
    const { isFetching } = this.props;

    if (isFetching) {
      return (<Preloader />);
    } else if (!isFetching && this.state.recipes.length === 0) {
      return (
        <p className="no-recipes">
          Sorry no recipes present at the moment
          <span role="img" aria-label="sad">üò¢  </span>, please try again later
          <span role="img" aria-label="please">üôè</span>
        </p>
      );
    }
    return (<TopRatedRecipeList recipes={this.state.recipes} />);
  }

  render() {
    return (
      <div className="page page__home">
        <header className="site-header">
          <ConnectedSiteNav currentUser={this.props.currentUser} />
          <section className="site-header__hero">
            <h2>Delicious recipes, just a search away...</h2>
            <input
              className="site-header__hero--input"
              type="text"
              onFocus={this.toggleSearch}
              placeholder="what would you like to cook?"
            />
          </section>
        </header>
        <main>
          <div className="container">
            <section className="page__home--top-rated">
              <header>
                <h1>Top Rated</h1>
              </header>
              {
                this.renderBody()
              }
              <footer>
                <ReactPaginate
                  previousLabel={<i className="material-icons">chevron_left</i>}
                  nextLabel={<i className="material-icons">chevron_right</i>}
                  pageCount={this.state.last}
                  marginPagesDisplayed={2}
                  pageClassName="waves-effect"
                  pageRangeDisplayed={4}
                  onPageChange={this.handlePageClick}
                  activeClassName="active"
                  containerClassName="pagination"
                />
              </footer>


            </section>
          </div>
        </main>
        {
          this.state.isSearchActive && <Search toggleSearch={this.toggleSearch} />
        }
        <SiteFooter />
      </div>
    );
  }
}

Home.propTypes = {
  dispatch: PropTypes.func.isRequired,
  isFetching: PropTypes.bool.isRequired,

  ...currentUserPropTypes
};

const mapStateToProps = ({
  topRecipes: { recipes, pagination },
  isFetching,
  auth: { currentUser }
}) => ({
  recipes,
  pagination,
  currentUser,
  isFetching
});

export { Home as PureHome };
export default connect(mapStateToProps)(Home);
