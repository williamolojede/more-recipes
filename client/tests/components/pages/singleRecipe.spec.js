import React from 'react';
import { PureSingleRecipe, mapStateToProps } from '../../../components/pages/SingleRecipe';
import userMock from '../../__mocks__/user.mock';
import recipesMock from '../../__mocks__/recipes.mock';
import notificationMock from '../../__mocks__/notification.mock';
import { initialState } from '../../../reducers/recipe';

function setup() {
  const props = {
    currentUser: userMock,
    dispatch: jest.fn(),
    recipe: recipesMock.recipes[1],
    match: {
      params: {
        id: '5',
      }
    },
    notification: notificationMock
  };
  const wrapper = shallow(<PureSingleRecipe {...props} />);

  return {
    wrapper
  };
}

describe('Single Recipe Page', () => {
  it('should render with right amount of elements', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render recipe not found view if there is an error', () => {
    const { wrapper } = setup();
    wrapper.setProps({ errorMessage: 'recipe not found' });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a preloader', () => {
    const { wrapper } = setup();
    wrapper.setProps({ recipe: initialState.recipe, isFetching: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should dispatch an action to vote for recipe when the method is called', () => {
    const { wrapper } = setup();
    const dispatchSpy = jest.spyOn(wrapper.instance().props, 'dispatch');
    wrapper.instance().vote('up');
    // why 2? because dispatch is called also in componentDidMount
    expect(dispatchSpy).toHaveBeenCalledTimes(2);
  });

  it('should dispatch an action to favorite a recipe when the method is called', () => {
    const { wrapper } = setup();
    const dispatchSpy = jest.spyOn(wrapper.instance().props, 'dispatch');
    wrapper.instance().componentWillUnmount();
    expect(dispatchSpy).toHaveBeenCalledTimes(2);
  });

  it('should dispatch an action to reset the single recipe store when the component is unmounted', () => {
    const { wrapper } = setup();
    const dispatchSpy = jest.spyOn(wrapper.instance().props, 'dispatch');
    wrapper.instance().favorite();
    expect(dispatchSpy).toHaveBeenCalledTimes(2);
  });

  describe('mapStateToProps', () => {
    it('should return an object with props required by component from store', () => {
      const store = {
        singleRecipe: {
          recipe: recipesMock.recipes[1],
          errorMessage: ''
        },
        auth: { currentUser: userMock },
        isFetching: false,
        notification: ''
      };

      const {
        recipe,
        errorMessage,
        currentUser,
        isFetching,
        notification
      } = mapStateToProps(store);

      expect(recipe.name).toBe(recipesMock.recipes[1].name);
      expect(errorMessage).toBe('');
      expect(currentUser.fullname).toBe(userMock.fullname);
      expect(isFetching).toBe(false);
      expect(notification).toBe('');
    });
  });
});
