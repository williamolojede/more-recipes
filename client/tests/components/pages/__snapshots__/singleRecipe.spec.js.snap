// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single Recipe Page should render a preloader 1`] = `
<div
  className="page page__recipe"
>
  <header
    className="site-header"
  >
    <Connect(SiteNav)
      currentUser={
        Object {
          "fullname": "John Doe",
          "id": 1,
        }
      }
    />
  </header>
  <Preloader />
  <Notification
    dispatch={[Function]}
    notification=""
  />
  <SiteFooter />
</div>
`;

exports[`Single Recipe Page should render recipe not found view if there is an error 1`] = `
<div
  className="page page__recipe"
>
  <header
    className="site-header"
  >
    <Connect(SiteNav)
      currentUser={
        Object {
          "fullname": "John Doe",
          "id": 1,
        }
      }
    />
  </header>
  <div
    className="recipe-not-found"
  >
    <h1>
      <span
        aria-label="sad"
        role="img"
      >
        ðŸ˜¢
      </span>
       Recipe Not Found
    </h1>
  </div>
  <Notification
    dispatch={[Function]}
    notification=""
  />
  <SiteFooter />
</div>
`;

exports[`Single Recipe Page should render with right amount of elements 1`] = `
<div
  className="page page__recipe"
>
  <header
    className="site-header"
  >
    <Connect(SiteNav)
      currentUser={
        Object {
          "fullname": "John Doe",
          "id": 1,
        }
      }
    />
  </header>
  <main>
    <div
      className="container"
    >
      <div
        className="row"
      >
        <div
          className="col s12 m5"
        >
          <img
            alt="Chef John's Spaghetti with Red Clam Sauce dfbefadb nifek fedckltefdkletfdmlefdleff eifdanlkref"
            src="http://images.media-allrecipes.com/userphotos/720x405/4478567.jpg"
          />
        </div>
        <div
          className="col s12 m7"
        >
          <h4
            className="recipe__name"
          >
            Chef John's Spaghetti with Red Clam Sauce dfbefadb nifek fedckltefdkletfdmlefdleff eifdanlkref
          </h4>
          <p
            className="recipe__description"
          >
            This is one of my 'go to' recipes and it takes only about 15 minutes from start to finish. The only prep work really is mashing the garlic. The sauce is great on any pasta, but I prefer angel hair or spaghetti.
          </p>
          <div
            className="row recipe__stats"
          >
            <RecipeStats
              favorite={[Function]}
              stats={
                Object {
                  "downVoteCount": 1,
                  "favoriteCount": 1,
                  "upVoteCount": 1,
                  "viewCount": 2,
                }
              }
              vote={[Function]}
            />
            <a
              className="write-review"
              href="#write-review"
            >
              Write Review
            </a>
          </div>
        </div>
      </div>
      <IngredientsList
        ingredients={
          Array [
            "2 tablespoons olive oil",
            "3 cloves garlic",
            "2 tablespoons capers",
            "1 tablespoon anchovy paste",
            "1 teaspoon red pepper flakes",
            "1 (24 ounce) jar tomato pasta sauce",
            "1/4 cup water",
            "1 cup good-quality crisp white wine",
            "2 (6.5 ounce) cans chopped clams, drained with juice reserved",
            "salt and pepper to taste",
            "1 pound dry spaghetti",
            "1/4 cup freshly grated Parmesan cheese",
            "Chopped fresh basil leaves or parsley",
          ]
        }
      />
      <InstructionsList
        instructions={
          Array [
            "Pour olive oil into cold skillet. Add garlic, capers, anchovy paste, and red pepper flakes. Place over medium heat and cook and stir until the oil is infused with the flavors of the mixture, about 5 minutes, taking care not to brown the garlic. Add tomato",
            "Bring a large pot of lightly salted water to a boil. Cook spaghetti in the boiling water, stirring occasionally until almost al dente, or about 1 minute less than directed on package. Drain. Return pasta to the pot off heat.",
            "Stir clams into the sauce. Pour sauce over pasta. Stir; cover the pot to allow it to finish cooking off heat and to allow the pasta will absorb some of the sauce, about 3 minutes. Stir in Parmesan cheese. Top with chopped fresh basil and/or parsley.",
          ]
        }
      />
      <Connect(Reviews)
        recipeId={5}
      />
    </div>
  </main>
  <Notification
    dispatch={[Function]}
    notification=""
  />
  <SiteFooter />
</div>
`;
